### PR TITLE
[DateInput/DateRangeInput] Commit typed value on 'Enter' key press

### LINKED ISSUE #1798: [DateTime] Fields don't "commit" the inputed date on 'Enter'
This is a feature request to support the use of "enter" when selecting dates from the Blueprint Date Picker.

### PR DESCRIPTION
**Pull request by cmslewis**
#### Checklist

- [x] Include tests

#### Changes proposed in this pull request:

- `DateInput` and `DateRangeInput` now save the typed value(s) on <kbd>Enter</kbd>.
    - They also move/transfer the focus state appropriately.

#### Reviewers should focus on:

- The tests don't corroborate my understanding of what should be happening. Maybe I just need to sleep on it. I've commented or omitted some checks in the meantime.
- Idea for followup: Revert to the old value on <kbd>Esc</kbd>? Thoughts?
- Does controlled mode look good?

#### Screenshot

_`DateInput`_

_`DateRangeInput`_

### CODE REVIEW

**Comment by llorca** [comment]
does this PR change any interaction?

+1 on reverting to previous value on `Esc`, if current value hasn't been submitted

**Comment by cmslewis** [comment]
@llorca no, it should be strictly additive.

File: packages/datetime/src/dateInput.tsx [hunk 1]
```diff
@@ -423,6 +425,15 @@ export class DateInput extends AbstractComponent<IDateInputProps, IDateInputStat
         this.safeInvokeInputProp("onBlur", e);
     };
 
+    private handleInputKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
+        if (e.which === Keys.ENTER) {
+            const nextValue = this.createMoment(this.state.valueString);
+            const nextDate = fromMomentToDate(nextValue);
+            this.handleDateChange(nextDate, true);
+            this.inputRef.blur();
+        }
```
Comment (cmslewis) [inline packages/datetime/src/dateInput.tsx]: Whoops, need to do `this.safeInvokeInputProp("onKeyDown", e);`.

File: packages/datetime/src/dateRangeInput.tsx [hunk 2]
```diff
@@ -542,9 +542,12 @@ export class DateRangeInput extends AbstractComponent<IDateRangeInputProps, IDat
     // - if focused in start field, Tab moves focus to end field
     // - if focused in end field, Shift+Tab moves focus to start field
     private handleInputKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
```
Comment (cmslewis) [inline packages/datetime/src/dateRangeInput.tsx]: Here too?

File: packages/datetime/src/dateInput.tsx [hunk 3]
```diff
@@ -423,6 +425,15 @@ export class DateInput extends AbstractComponent<IDateInputProps, IDateInputStat
         this.safeInvokeInputProp("onBlur", e);
     };
 
+    private handleInputKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
+        if (e.which === Keys.ENTER) {
+            const nextValue = this.createMoment(this.state.valueString);
+            const nextDate = fromMomentToDate(nextValue);
+            this.handleDateChange(nextDate, true);
+            this.inputRef.blur();
```
Comment (giladgray) [inline packages/datetime/src/dateInput.tsx]: 🤔 this `.blur()` feels intuitively wrong, as it kills the focus flow: I can't hit `enter` and then `tab` to move to the next form input.

can we do everything else but simply not blur? same goes for DRI.

Comment (cmslewis) [reply packages/datetime/src/dateInput.tsx]: I buy that.

**Comment by cmslewis** [comment]
@giladgray is this good to merge?

File: packages/datetime/src/dateInput.tsx (hunk 1 above)
Comment (giladgray) [reply packages/datetime/src/dateInput.tsx]: @cmslewis is this comment still relevant?

**Comment by giladgray** [comment]
@cmslewis pressing `enter` in `DateInput` does not expand to full date. works great for `DRI` though.

**Comment by adidahiya** [comment]
> pressing enter in DateInput does not expand to full date

seems fine to address in a follow up. current behavior seems fine to me. @erinmarchand @LindsayWard @dalessi can you verify in the "documentation" link above?

**Comment by giladgray** [comment]
wait @adidahiya how is this fine to merge? it literally doesn't work properly in one of the title use cases.

**Comment by adidahiya** [comment]
It seems to address the original intent of the filed issue though. Once you type something in DateInput, you get it committed when you blur out of the input (the popover gets closed, which suggests to the user that something worked). Seems reasonable in most cases, and we can make further adjustments if we get more feedback. But I'm also cool with waiting to hear from the people tagged in my last comment ^

**Comment by llorca** [comment]
Check out this GIF:
screen recording google chrome
What's good: in both DI and DRI, the date picker activates with the date typed in the field
What's not good: DRI updates the input value on `enter`, DI doesn't. Instead, DI updates the input value on blur.

Not a fan of that inconsistency, as @giladgray pointed out.

**Comment by cmslewis** [comment]
@llorca @adidahiya @giladgray Sorry for the delay. Should be good to go now.

**Comment by adidahiya** [comment]
nice, lgtm 👍

**Pull request merged**